sql/table.go

```diff
@@ -473,3 +473,65 @@ func convertDatum(col ColumnDescriptor, val parser.Datum) (interface{}, error) {
 	return nil, fmt.Errorf("value type %s doesn't match type %s of column %q",
 		val.Type(), col.Type.Kind, col.Name)
 }
+
+// unmarshalColumnValue decodes the value from a key-value pair using the type
+// expected by the column. An error is returned if the value's type does not
+// match the column's type.
+func unmarshalColumnValue(kind ColumnType_Kind, value *proto.Value) (parser.Datum, error) {
+	if value == nil {
+		return parser.DNull, nil
+	}
+
+	switch kind {
+	case ColumnType_BOOL:
+		v, err := value.GetInt()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DBool(v != 0), nil
+	case ColumnType_INT:
+		v, err := value.GetInt()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DInt(v), nil
+	case ColumnType_FLOAT:
+		v, err := value.GetFloat()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DFloat(v), nil
+	case ColumnType_STRING:
+		v, err := value.GetBytesChecked()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DString(v), nil
+	case ColumnType_BYTES:
+		v, err := value.GetBytesChecked()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DBytes(v), nil
+	case ColumnType_DATE:
+		v, err := value.GetTime()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DDate{Time: v}, nil
+	case ColumnType_TIMESTAMP:
+		v, err := value.GetTime()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DTimestamp{Time: v}, nil
+	case ColumnType_INTERVAL:
+		v, err := value.GetInt()
+		if err != nil {
+			return nil, err
+		}
+		return parser.DInterval{Duration: time.Duration(v)}, nil
```

Comment:
Nit: I intentionally suggested `util.Errorf` here because this error would indicate we (the cockroach devs) did something wrong, rather than the user.